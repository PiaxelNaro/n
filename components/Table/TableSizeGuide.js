// @flow
/* eslint react/no-array-index-key: 0 */

import React from 'react';
import styled from 'styled-components';
import { type Column } from './utils';

const TableGuide = styled.tbody`
    height: 0;

    .table-hide {
        border: none;
        padding: 0;
    }
`;

const SIZE_MAP = {
    auto: {
        weight: 0,
        minWidth: 'auto',
    },
    extraSmall: {
        weight: 0.6,
        minWidth: '60px',
    },
    small: {
        weight: 0.8,
        minWidth: '80px',
    },
    medium: {
        weight: 1.2,
        minWidth: '120px',
    },
    large: {
        weight: 2.6,
        minWidth: '260px',
    },
    extraLarge: {
        weight: 4,
        minWidth: '400px',
    },
};

const TableSizeGuide = ({ columns }: { columns: Array<Column> }) => {
    const sizes = columns.reduce((m, column) => {
        if (column.subColumns.length) {
            return m.concat(...column.subColumns.map(({ size }) => size));
        }
        return m.concat(column.size);
    }, []);
    const totalWeight = sizes.reduce((m, size) => m + SIZE_MAP[size].weight, 0);

    return (
        <TableGuide>
            <tr className="table-hide">
                {sizes.map((size, i) => {
                    const perc = SIZE_MAP[size].weight / totalWeight * 100;
                    return (
                        <td
                            className="table-hide"
                            key={i}
                            style={{
                                width: perc > 0 ? `${perc}%` : 'auto',
                                minWidth: SIZE_MAP[size].minWidth,
                            }}
                        />
                    );
                })}
            </tr>
        </TableGuide>
    );
};

export default TableSizeGuide;
