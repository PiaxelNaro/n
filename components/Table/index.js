// @flow

import React from 'react';
import styled, { css } from 'styled-components';
import isEqual from 'lodash/isEqual';
import identity from 'lodash/identity';
import noop from 'lodash/noop';
import classnames from 'classnames';
import { scrollAggregator, resizeAggregator } from '../../modules/eventAggregator';
import {
    type LooseColumn,
    type Column,
    createSizeMatrix,
    normalizeColumn,
} from './utils';
import TableSizeGuide from './TableSizeGuide';
import TableHeader, { type SortingRule } from './TableHeader';

type FormatterMap = { [string]: (any) => any };
type TableSizeMatrix = {
    width: string,
    scrollerWidth: string,
    rows: Array<{ height: string, cells: Array<{ width: string }> }>,
};

const borderStyle = '1px solid #D9DCDF';
const StyledTable = styled.table`
    min-width: 100%;
    border-collapse: collapse;
    border-spacing: 0;
    font-size: 13px;
    text-align: left;
    color: #666;

    td,
    th {
        padding: 6px;
        box-sizing: border-box;
        border-bottom: ${borderStyle};
    }

    td + td {
        border-left: ${borderStyle};
    }

    ${props =>
        props.cellAlignment.map(
            (align, index) => css`
                td:nth-child(${index + 1}) {
                    text-align: ${align};
                }
            `
        )}

    .combined-column {
        border-left: none;
    }
`;

const TableBody = styled.tbody`
    background: #fff;
`;

const TableWrapper = styled.div`
    position: relative;
    width: 100%;
`;

const TableScroll = styled.div`
    display: block;
    width: 100%;
    overflow: hidden;
    overflow-x: auto;
`;

const FixedHeader = styled.div`
    position: absolute;
    top: 0;
    &.is-fixed {
        position: fixed;
    }
`;

const ScrollTableWrapper = styled.div`
    overflow: hidden;
`;

const FixedSidebar = styled.div`
    position: absolute;
    top: 0;
    left: 0;
    bottom: 0;
    table {
        border-right: ${borderStyle};
        border-bottom: none;
    }
`;

const getKeyGroupsFromColumns = (columns: Array<Column>): Array<Array<string>> =>
    columns.reduce(
        (m, col) => m.concat(col.subColumns.map(({ keys }) => keys), [col.keys]),
        []
    );

const getFormatter = (formatters: FormatterMap, key: string) =>
    formatters[key] || identity;

function getCellAlignment(columns: Array<Column>): Array<string> {
    return columns.reduce(
        (memory, column) =>
            memory.concat(
                column.align,
                column.subColumns.reduce(
                    (subMemory, subCol) => subMemory.concat(subCol.align),
                    []
                )
            ),
        []
    );
}

const AATableRow = ({
    item,
    keyGroups,
    height,
    formatters = {},
}: {
    item: Object,
    keyGroups: Array<Array<string>>,
    height?: ?string,
    formatters?: FormatterMap,
}) =>
    <tr style={{ height }}>
        {keyGroups.map(keys =>
            keys.map((key, keyIndex) =>
                <td key={key} className={classnames({ 'combined-column': keyIndex > 0 })}>
                    {getFormatter(formatters, key)(item[key])}
                </td>
            )
        )}
    </tr>;

class BaseTable extends React.PureComponent {
    props: {
        columns: Array<LooseColumn>,
        items: Array<Object>,
        onSort?: (string, string) => any,
        sortedBy?: SortingRule,
        formatters?: FormatterMap,
        innerRef: (HTMLElement | void) => any,
        onResize: TableSizeMatrix => any,
        onScroll: number => any,
    };

    tableScroller: HTMLElement;
    tableRef: HTMLElement;
    unsubscribeTableScroll: Function;
    unsubscribeResize: Function;
    generateSizeMatrix: Function;
    handleScroll: Function;

    static defaultProps = {
        innerRef: noop,
        onResize: noop,
        onScroll: noop,
    };

    constructor() {
        super();

        this.unsubscribeTableScroll = this.unsubscribeTableScroll.bind(this);
        this.generateSizeMatrix = this.generateSizeMatrix.bind(this);
        this.handleScroll = this.handleScroll.bind(this);
    }

    componentDidMount() {
        this.unsubscribeResize = resizeAggregator(this.generateSizeMatrix);

        this.unsubscribeTableScroll();
        if (this.tableScroller) {
            this.tableScroller.addEventListener('scroll', this.handleScroll);
        }
    }

    componentDidUpdate() {
        this.generateSizeMatrix();
    }

    componentWillUnmount() {
        this.unsubscribeResize();
        this.unsubscribeTableScroll();
    }

    generateSizeMatrix() {
        const sizeMatrix = createSizeMatrix(this.tableScroller);
        this.props.onResize(sizeMatrix);
    }

    handleScroll() {
        if (this.tableScroller) {
            const left = this.tableScroller.scrollLeft;
            this.props.onScroll(left);
        }
    }

    unsubscribeTableScroll() {
        if (this.tableScroller) {
            this.tableScroller.removeEventListener('scroll', this.handleScroll);
        }
    }

    render() {
        const { items, onSort, sortedBy, formatters } = this.props;
        const columns = this.props.columns.map(normalizeColumn);
        const cellAlignment = getCellAlignment(columns);
        const keyGroups = getKeyGroupsFromColumns(columns);

        return (
            <TableScroll
                className="TableScroll"
                innerRef={tableScroller => (this.tableScroller = tableScroller)}
            >
                <StyledTable
                    cellAlignment={cellAlignment}
                    innerRef={tableRef => {
                        this.tableRef = tableRef;
                        this.props.innerRef(tableRef);
                    }}
                >
                    <TableHeader columns={columns} onSort={onSort} sortedBy={sortedBy} />
                    <TableBody>
                        {items.map(item =>
                            <AATableRow
                                key={item.id}
                                item={item}
                                keyGroups={keyGroups}
                                formatters={formatters}
                            />
                        )}
                    </TableBody>
                    <TableSizeGuide columns={columns} />
                </StyledTable>
            </TableScroll>
        );
    }
}

class AATable extends React.Component {
    props: {
        fixedColumns?: number,
        columns: Array<LooseColumn>,
        items: Array<Object>,
        onSort?: (string, string) => any,
        sortedBy?: SortingRule,
        formatters?: FormatterMap,
        innerRef: (HTMLElement | void) => any,
    };

    state: {
        sizeMatrix?: TableSizeMatrix,
    };

    static defaultProps = {
        fixedColumns: 0,
        innerRef: noop,
    };

    documentScrollWatcher: Function;
    tableScrollWatcher: Function;
    tableInnerRef: Function;
    onResize: Function;

    unsubscribeScroll: Function;
    unsubscribeResize: Function;

    fixedHeaderRef: HTMLElement;
    tableRef: HTMLElement;

    constructor() {
        super();

        this.state = {};

        this.documentScrollWatcher = this.documentScrollWatcher.bind(this);
        this.tableScrollWatcher = this.tableScrollWatcher.bind(this);
        this.tableInnerRef = this.tableInnerRef.bind(this);
        this.onResize = this.onResize.bind(this);
    }

    onResize(sizeMatrix: TableSizeMatrix) {
        // Only update state if the matrix changed. This is to prevent an infinite loop we'd
        // have otherwise.
        if (!isEqual(sizeMatrix, this.state.sizeMatrix)) {
            this.setState({ sizeMatrix });
        }
    }

    documentScrollWatcher() {
        if (this.fixedHeaderRef && this.tableRef) {
            const rect = this.tableRef.getBoundingClientRect();
            if (rect.top < 0 && rect.bottom > 0) {
                this.fixedHeaderRef.classList.add('is-fixed');
            } else {
                this.fixedHeaderRef.classList.remove('is-fixed');
            }
        }
    }

    tableScrollWatcher(left: number) {
        if (
            this.fixedHeaderRef &&
            this.fixedHeaderRef.firstChild instanceof HTMLElement
        ) {
            this.fixedHeaderRef.firstChild.scrollLeft = left;
        }
    }

    tableInnerRef(tableRef: HTMLElement) {
        this.tableRef = tableRef;
        this.props.innerRef(tableRef);
    }

    componentDidMount() {
        this.unsubscribeScroll = scrollAggregator(this.documentScrollWatcher);
    }

    componentWillUnmount() {
        this.unsubscribeScroll();
    }

    render() {
        const { items, fixedColumns, onSort, sortedBy, formatters } = this.props;
        const columns = this.props.columns.map(normalizeColumn);
        const cellAlignment = getCellAlignment(columns);

        let fixedLayout;
        const sizeMatrix = this.state.sizeMatrix;
        if (sizeMatrix) {
            const columnsToFix = columns.slice(0, fixedColumns);
            const fixedKeyGroups = getKeyGroupsFromColumns(columnsToFix);
            const sidebarHeader = (
                <TableHeader
                    columns={columnsToFix}
                    sizeMatrix={sizeMatrix}
                    onSort={onSort}
                    sortedBy={sortedBy}
                />
            );
            fixedLayout = [
                <FixedSidebar key="FixedSidebar">
                    <StyledTable cellAlignment={cellAlignment}>
                        {sidebarHeader}
                        <TableBody>
                            {items.map((item, i) =>
                                <AATableRow
                                    key={item.id}
                                    item={item}
                                    keyGroups={fixedKeyGroups}
                                    formatters={formatters}
                                    height={
                                        sizeMatrix.rows[i + 2]
                                            ? sizeMatrix.rows[i + 2].height
                                            : null
                                    }
                                />
                            )}
                        </TableBody>
                    </StyledTable>
                </FixedSidebar>,
                <FixedHeader
                    innerRef={fixedHeaderRef => (this.fixedHeaderRef = fixedHeaderRef)}
                    style={{ width: sizeMatrix.scrollerWidth }}
                    key="FixedHeader"
                >
                    <ScrollTableWrapper>
                        <StyledTable
                            cellAlignment={cellAlignment}
                            style={{ width: sizeMatrix.width }}
                        >
                            <TableHeader
                                columns={columns}
                                sizeMatrix={sizeMatrix}
                                onSort={onSort}
                                sortedBy={sortedBy}
                            />
                        </StyledTable>
                        <FixedSidebar>
                            <StyledTable cellAlignment={cellAlignment}>
                                {sidebarHeader}
                            </StyledTable>
                        </FixedSidebar>
                    </ScrollTableWrapper>
                </FixedHeader>,
            ];
        }

        return (
            <TableWrapper className="TableWrapper">
                <BaseTable
                    innerRef={this.tableInnerRef}
                    onScroll={this.tableScrollWatcher}
                    onResize={this.onResize}
                    columns={this.props.columns}
                    items={this.props.items}
                    onSort={this.props.onSort}
                    sortedBy={this.props.sortedBy}
                    formatters={this.props.formatters}
                />
                {fixedLayout}
            </TableWrapper>
        );
    }
}

export default AATable;
export { AATableRow, StyledTable, BaseTable };
