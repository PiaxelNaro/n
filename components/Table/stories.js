import React from 'react';
import { storiesOf, action } from '@kadira/storybook';
import { withKnobs, number } from '@kadira/storybook-addon-knobs';
import Table from '.';
import tableData from '../../data/table-1';
import apps from '../../data/apps';

const injectRandomApp = items =>
    items.map(data => {
        const installedApps = [];
        while (installedApps.length < 10 && Math.random() < 0.7) {
            installedApps.push(apps[Math.floor(Math.random() * apps.length)]);
        }

        let appDiff = Math.floor(Math.random() * 10) - installedApps.length;
        if (appDiff > 0) {
            appDiff = `+${appDiff}`;
        } else if (appDiff === 0) {
            appDiff = '=';
        }

        return {
            ...data,
            installed_apps_icons: installedApps.map(({ icon }) => icon),
            installed_apps_count: installedApps.length,
            installed_apps_count_diff: appDiff,
        };
    });

const demoData = injectRandomApp(tableData);

const columns = [
    {
        label: 'First Name',
        tooltip: 'Lorem ipsum dolor sit amet, consectetur adipisicing elit. Alias, officiis.',
        key: 'first_name',
        isSortable: true,
        align: 'right',
    },
    {
        label: 'Last Name',
        key: 'last_name',
        isSortable: true,
        align: 'center',
    },
    {
        label: 'Email',
        key: 'email',
        isSortable: true,
        align: 'right',
    },
    {
        label: 'Installed Apps',
        subColumns: [
            {
                label: 'Apps',
                tooltip: 'Lorem ipsum dolor sit amet, consectetur adipisicing elit.',
                key: 'installed_apps_icons',
                isSortable: true,
                size: 'large',
                align: 'right',
            },
            {
                label: 'Count',
                tooltip: 'Lorem ipsum dolor sit amet, consectetur.',
                keys: ['installed_apps_count', 'installed_apps_count_diff'],
                isSortable: true,
                size: ['medium', 'auto'],
                align: ['center', 'left'],
            },
        ],
    },
    {
        label: 'Country',
        keys: ['country', 'ip_address'],
        isSortable: true,
        size: 'auto',
    },
];
const simpleColumns = [
    {
        label: 'First Name',
        tooltip: 'Lorem ipsum dolor.',
        key: 'first_name',
        isSortable: true,
    },
    {
        label: 'Last Name',
        tooltip: 'Lorem ipsum dolor sit.',
        key: 'last_name',
        isSortable: true,
    },
    {
        label: 'Email',
        key: 'email',
        isSortable: true,
    },
    {
        label: 'Installed Apps',
        key: 'installed_apps_icons',
    },
    {
        label: 'Country',
        keys: ['country', 'ip_address'],
        isSortable: true,
        size: 'auto',
    },
];

const formatters = {
    installed_apps_icons: icons =>
        icons.map(icon => (
            <img key={Math.random()} src={icon} height="40px" width="40px" alt="" />
        )),
};

class Demo extends React.Component {
    onSort: Function;
    props: {
        simple: Boolean,
    };

    constructor() {
        super();

        this.state = {
            sortBy: {
                key: 'first_name',
                dir: 'asc',
            },
        };

        this.onSort = this.onSort.bind(this);
    }

    onSort(key, dir) {
        action('sort by')(key, dir);
        this.setState({
            sortBy: { key, dir },
        });
    }

    render() {
        return (
            <div>
                <div style={{ height: '50px' }} />
                <Table
                    fixedColumns={number('Number of fixed columns', 1)}
                    items={demoData}
                    onSort={this.onSort}
                    sortedBy={this.state.sortBy}
                    columns={this.props.simple ? simpleColumns : columns}
                    formatters={formatters}
                />
                <div style={{ height: '3000px' }} />
            </div>
        );
    }
}

storiesOf('Table', module)
    .addDecorator(withKnobs)
    .add('Sticky header and columns', () => <Demo />)
    .add('Simple header demo', () => <Demo simple />);
