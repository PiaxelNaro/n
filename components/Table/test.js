// @flow
import React from 'react';
import { shallow } from 'enzyme';
import Table, { AATableRow, StyledTable, BaseTable } from '.';

describe('<AATableRow />', () => {
    it('render table row with sub columns and combined columns', () => {
        const keyGroups = [['a'], ['b', 'c'], ['d'], ['e', 'f']];
        const item = {
            a: '#1',
            b: '#2',
            c: '#3',
            d: '#4',
            e: '#5',
            f: '#6',
        };

        expect(
            shallow(<AATableRow keyGroups={keyGroups} item={item} height="18px" />)
        ).toMatchSnapshot();
    });

    it('allows passing formatters to render cells', () => {
        const aFormatter = jest.fn(() => 'b');
        const keyGroups = [['a']];
        const item = { a: 'a' };
        const formatters = { a: aFormatter };

        const row = shallow(
            <AATableRow
                keyGroups={keyGroups}
                item={item}
                formatters={formatters}
                height="18px"
            />
        );
        expect(aFormatter).toHaveBeenCalledWith('a');
        expect(row.find('td').text()).toEqual('b');
    });
});

describe('<BaseTable />', () => {
    // Non-normalized columns, the Table component handles that itself.
    const columns = [
        { key: 'a', label: 'First' },
        { keys: ['b', 'c'], label: 'Second' },
        {
            label: 'Multi col',
            subColumns: [{ key: 'd', label: 'Foo' }, { keys: ['e', 'f'], label: 'Bar' }],
        },
    ];

    const items = [
        {
            a: '1a',
            b: '2b',
            c: '3c',
            d: '4d',
            e: '5e',
            f: '6f',
        },
        {
            a: '1a',
            b: '2b',
            c: '3c',
            d: '4d',
            e: '5e',
            f: '6f',
        },
    ];

    it('renders base table', () => {
        const sortSpy = jest.fn();
        const table = shallow(
            <BaseTable
                columns={columns}
                items={items}
                onSort={sortSpy}
                sortedBy={{ key: 'a', dir: 'asc' }}
                formatters={{}}
            />
        );
        expect(table).toMatchSnapshot();

        expect(table.find('TableHeader').prop('onSort')).toBe(sortSpy);
    });

    it('triggers scroll events', () => {
        global.requestAnimationFrame = jest.fn();
        const scrollSpy = jest.fn();
        const table = shallow(
            <BaseTable columns={columns} items={items} onScroll={scrollSpy} />
        );
        // Emulate table mounting without tableScroller ref
        const inst: any = table.instance();
        table.find('Table__TableScroll').prop('innerRef')(undefined);
        inst.componentDidMount();
        table.update();

        // Emulate table mounting with scroll ref binding
        const tableScrollRef: any = {
            querySelector: jest.fn(() => null),
            addEventListener: jest.fn(),
            removeEventListener: jest.fn(),
        };
        table.find('Table__TableScroll').prop('innerRef')(tableScrollRef);
        inst.componentDidMount();
        table.update();

        // Manually trigger a horizontal scroll event on the table scroll
        tableScrollRef.scrollLeft = 10;
        tableScrollRef.addEventListener.mock.calls[0][1]();
        expect(scrollSpy).toHaveBeenCalledWith(10);

        // An async scroll event might be triggered after the BaseTable is unmounted, test edge
        // case handling
        table.find('Table__TableScroll').prop('innerRef')(undefined);
        tableScrollRef.addEventListener.mock.calls[0][1]();
        inst.componentWillUnmount();
    });

    it('triggers resize events', () => {
        global.requestAnimationFrame = jest.fn();
        const resizeSpy = jest.fn();
        const table = shallow(
            <BaseTable columns={columns} items={items} onResize={resizeSpy} />
        );

        const firstChild = {
            tagName: 'TH',
            offsetWidth: 45,
            getBoundingClientRect: () => ({
                height: 18,
            }),
        };
        const tableRef: any = {
            offsetWidth: 800,
            querySelectorAll: jest.fn(() => [
                {
                    parentNode: {
                        tagName: 'THEAD',
                    },
                    children: [
                        firstChild,
                        { offsetWidth: 60 },
                        { offsetWidth: 80 },
                        { offsetWidth: 45 },
                        { offsetWidth: 60 },
                        { offsetWidth: 80 },
                    ],
                    getBoundingClientRect: () => ({
                        height: 18,
                    }),
                },
                {
                    parentNode: {
                        tagName: 'THEAD',
                    },
                    children: [firstChild],
                    getBoundingClientRect: () => ({
                        height: 18,
                    }),
                },
                {
                    parentNode: {
                        tagName: 'TBODY',
                    },
                    children: [
                        {
                            tagName: 'TD',
                            getBoundingClientRect: () => ({
                                height: 20,
                            }),
                        },
                    ],
                    getBoundingClientRect: () => ({
                        height: 20,
                    }),
                },
            ]),
        };
        table.find('Table__StyledTable').prop('innerRef')(tableRef);

        const tableScrollerRef: any = {
            querySelector: jest.fn(() => tableRef),
            getBoundingClientRect: () => ({
                width: 600,
            }),
            addEventListener: jest.fn(),
            removeEventListener: jest.fn(),
        };
        table.find('Table__TableScroll').prop('innerRef')(tableScrollerRef);

        // Emulate table mounting
        const inst: any = table.instance();
        inst.componentDidMount();
        inst.componentDidUpdate();
        table.update();

        expect(resizeSpy).toHaveBeenCalledWith({
            rows: [
                {
                    cells: [
                        { width: '45px' },
                        { width: '60px' },
                        { width: '80px' },
                        { width: '45px' },
                        { width: '60px' },
                        { width: '80px' },
                    ],
                    height: '18px',
                },
                {
                    cells: [{ width: '45px' }],
                    height: '18px',
                },
                { cells: [], height: '20px' },
            ],
            width: '800px',
            scrollerWidth: '600px',
        });

        inst.componentWillUnmount();
    });
});

describe('<Table />', () => {
    // Non-normalized columns, the Table component handles that itself.
    const columns = [
        { key: 'a', label: 'First' },
        { keys: ['b', 'c'], label: 'Second' },
        {
            label: 'Multi col',
            subColumns: [{ key: 'd', label: 'Foo' }, { keys: ['e', 'f'], label: 'Bar' }],
        },
    ];

    const items = [
        {
            a: '1a',
            b: '2b',
            c: '3c',
            d: '4d',
            e: '5e',
            f: '6f',
        },
        {
            a: '1a',
            b: '2b',
            c: '3c',
            d: '4d',
            e: '5e',
            f: '6f',
        },
    ];

    const sizeMatrix = {
        rows: [
            {
                cells: [
                    { width: '45px' },
                    { width: '60px' },
                    { width: '80px' },
                    { width: '45px' },
                    { width: '60px' },
                    { width: '80px' },
                ],
                height: '18px',
            },
            { cells: [], height: '18px' },
            { cells: [], height: '18px' },
        ],
        width: '800px',
    };

    it('render the styled table element', () => {
        const table = shallow(
            <StyledTable
                cellAlignment={['left', 'left', 'left', 'left', 'left', 'left']}
            />
        );
        expect(table).toMatchSnapshot();
    });

    it('renders fixed layout over the base table size matrix', () => {
        const table = shallow(<Table fixedColumns={1} columns={columns} items={items} />);
        table.find('BaseTable').prop('onResize')(sizeMatrix);

        // After BaseTable mount, the table will render the fixed layout.
        expect(table.find('Table__FixedSidebar').length).toEqual(2);
        expect(table.find('Table__FixedHeader').length).toEqual(1);
        expect(table).toMatchSnapshot();

        // The table won't re-render if the size matrix stayed the same.
        table.find('BaseTable').prop('onResize')(sizeMatrix);
    });

    it('sync the table scroll position and the fixed header', () => {
        const table = shallow(<Table fixedColumns={1} columns={columns} items={items} />);
        table.find('BaseTable').prop('onResize')(sizeMatrix);

        // Setup fake reference to the fixed header
        const fixedHeaderRef = {
            firstChild: global.document.createElement('div'),
        };
        table.find('Table__FixedHeader').prop('innerRef')(fixedHeaderRef);

        // Manually trigger a horizontal scroll event on the table scroll
        table.find('BaseTable').prop('onScroll')(10);
        expect(fixedHeaderRef.firstChild.scrollLeft).toEqual(10);

        table.find('BaseTable').prop('onScroll')(30);
        expect(fixedHeaderRef.firstChild.scrollLeft).toEqual(30);
    });

    it('fixed header to match page vertical scroll position', () => {
        const document = global.document;
        global.requestAnimationFrame = jest.fn();
        document.addEventListener = jest.fn();
        const table = shallow(<Table fixedColumns={1} columns={columns} items={items} />);

        const tableRef: any = {};
        table.find('BaseTable').prop('innerRef')(tableRef);
        table.find('BaseTable').prop('onResize')(sizeMatrix);
        const inst: any = table.instance();
        inst.componentDidMount();
        table.update();

        const fixedHeaderRef = {
            firstChild: document.createElement('div'),
            classList: {
                add: jest.fn(),
                remove: jest.fn(),
            },
        };
        table.find('Table__FixedHeader').prop('innerRef')(fixedHeaderRef);

        // Manually trigger a document scroll to exercise the sticky header
        tableRef.getBoundingClientRect = jest.fn(() => ({
            top: -20,
            bottom: 120,
        }));
        document.body.scrollTop = 20;
        document.addEventListener.mock.calls[0][1]();
        global.requestAnimationFrame.mock.calls[0][0]();
        expect(fixedHeaderRef.classList.add).toHaveBeenCalledWith('is-fixed');

        tableRef.getBoundingClientRect = jest.fn(() => ({
            top: 20,
            bottom: 120,
        }));
        document.body.scrollTop = 21;
        document.addEventListener.mock.calls[0][1]();
        global.requestAnimationFrame.mock.calls[1][0]();
        expect(fixedHeaderRef.classList.remove.mock.calls[0][0]).toEqual('is-fixed');

        tableRef.getBoundingClientRect = jest.fn(() => ({
            top: -20,
            bottom: -10,
        }));
        document.body.scrollTop = 22;
        document.addEventListener.mock.calls[0][1]();
        global.requestAnimationFrame.mock.calls[2][0]();
        expect(fixedHeaderRef.classList.remove.mock.calls[1][0]).toEqual('is-fixed');

        // Test unmounting
        inst.componentWillUnmount();
    });

    it('ensure type safety of view DOM', () => {
        const table = shallow(<Table fixedColumns={1} columns={columns} items={items} />);
        const inst: any = table.instance();

        inst.documentScrollWatcher();
        inst.tableScrollWatcher(0);
        inst.componentDidMount();
        inst.componentWillUnmount();
    });
});
