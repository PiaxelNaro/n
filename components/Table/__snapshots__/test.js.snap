// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AATableRow /> render table row with sub columns and combined columns 1`] = `
<tr
  style={
    Object {
      "height": "18px",
    }
  }
>
  <td
    className=""
  >
    #1
  </td>
  <td
    className=""
  >
    #2
  </td>
  <td
    className="combined-column"
  >
    #3
  </td>
  <td
    className=""
  >
    #4
  </td>
  <td
    className=""
  >
    #5
  </td>
  <td
    className="combined-column"
  >
    #6
  </td>
</tr>
`;

exports[`<BaseTable /> renders base table 1`] = `
<Table__TableScroll
  className="TableScroll"
  innerRef={[Function]}
>
  <Table__StyledTable
    cellAlignment={
      Array [
        "left",
        "left",
        "left",
        "left",
        "left",
        "left",
      ]
    }
    innerRef={[Function]}
  >
    <TableHeader
      columns={
        Array [
          Object {
            "align": Array [
              "left",
            ],
            "headerAlign": "left",
            "isSortable": false,
            "key": "a",
            "keys": Array [
              "a",
            ],
            "label": "First",
            "size": Array [
              "medium",
            ],
            "subColumns": Array [],
            "tooltip": undefined,
          },
          Object {
            "align": Array [
              "left",
              "left",
            ],
            "headerAlign": "left",
            "isSortable": false,
            "key": "b|c",
            "keys": Array [
              "b",
              "c",
            ],
            "label": "Second",
            "size": Array [
              "medium",
              "medium",
            ],
            "subColumns": Array [],
            "tooltip": undefined,
          },
          Object {
            "align": Array [],
            "headerAlign": "left",
            "isSortable": false,
            "key": "d|e|f",
            "keys": Array [],
            "label": "Multi col",
            "size": Array [],
            "subColumns": Array [
              Object {
                "align": Array [
                  "left",
                ],
                "headerAlign": "left",
                "isSortable": false,
                "key": "d",
                "keys": Array [
                  "d",
                ],
                "label": "Foo",
                "size": Array [
                  "medium",
                ],
                "subColumns": Array [],
                "tooltip": undefined,
              },
              Object {
                "align": Array [
                  "left",
                  "left",
                ],
                "headerAlign": "left",
                "isSortable": false,
                "key": "e|f",
                "keys": Array [
                  "e",
                  "f",
                ],
                "label": "Bar",
                "size": Array [
                  "medium",
                  "medium",
                ],
                "subColumns": Array [],
                "tooltip": undefined,
              },
            ],
            "tooltip": undefined,
          },
        ]
      }
      onSort={[Function]}
      sortedBy={
        Object {
          "dir": "asc",
          "key": "a",
        }
      }
    />
    <Table__TableBody>
      <AATableRow
        formatters={Object {}}
        item={
          Object {
            "a": "1a",
            "b": "2b",
            "c": "3c",
            "d": "4d",
            "e": "5e",
            "f": "6f",
          }
        }
        keyGroups={
          Array [
            Array [
              "a",
            ],
            Array [
              "b",
              "c",
            ],
            Array [
              "d",
            ],
            Array [
              "e",
              "f",
            ],
            Array [],
          ]
        }
      />
      <AATableRow
        formatters={Object {}}
        item={
          Object {
            "a": "1a",
            "b": "2b",
            "c": "3c",
            "d": "4d",
            "e": "5e",
            "f": "6f",
          }
        }
        keyGroups={
          Array [
            Array [
              "a",
            ],
            Array [
              "b",
              "c",
            ],
            Array [
              "d",
            ],
            Array [
              "e",
              "f",
            ],
            Array [],
          ]
        }
      />
    </Table__TableBody>
    <TableSizeGuide
      columns={
        Array [
          Object {
            "align": Array [
              "left",
            ],
            "headerAlign": "left",
            "isSortable": false,
            "key": "a",
            "keys": Array [
              "a",
            ],
            "label": "First",
            "size": Array [
              "medium",
            ],
            "subColumns": Array [],
            "tooltip": undefined,
          },
          Object {
            "align": Array [
              "left",
              "left",
            ],
            "headerAlign": "left",
            "isSortable": false,
            "key": "b|c",
            "keys": Array [
              "b",
              "c",
            ],
            "label": "Second",
            "size": Array [
              "medium",
              "medium",
            ],
            "subColumns": Array [],
            "tooltip": undefined,
          },
          Object {
            "align": Array [],
            "headerAlign": "left",
            "isSortable": false,
            "key": "d|e|f",
            "keys": Array [],
            "label": "Multi col",
            "size": Array [],
            "subColumns": Array [
              Object {
                "align": Array [
                  "left",
                ],
                "headerAlign": "left",
                "isSortable": false,
                "key": "d",
                "keys": Array [
                  "d",
                ],
                "label": "Foo",
                "size": Array [
                  "medium",
                ],
                "subColumns": Array [],
                "tooltip": undefined,
              },
              Object {
                "align": Array [
                  "left",
                  "left",
                ],
                "headerAlign": "left",
                "isSortable": false,
                "key": "e|f",
                "keys": Array [
                  "e",
                  "f",
                ],
                "label": "Bar",
                "size": Array [
                  "medium",
                  "medium",
                ],
                "subColumns": Array [],
                "tooltip": undefined,
              },
            ],
            "tooltip": undefined,
          },
        ]
      }
    />
  </Table__StyledTable>
</Table__TableScroll>
`;

exports[`<Table /> render the styled table element 1`] = `
<table
  className="Table__StyledTable-d861wc-0 ewdWPX"
/>
`;

exports[`<Table /> renders fixed layout over the base table size matrix 1`] = `
<Table__TableWrapper
  className="TableWrapper"
>
  <BaseTable
    columns={
      Array [
        Object {
          "key": "a",
          "label": "First",
        },
        Object {
          "keys": Array [
            "b",
            "c",
          ],
          "label": "Second",
        },
        Object {
          "label": "Multi col",
          "subColumns": Array [
            Object {
              "key": "d",
              "label": "Foo",
            },
            Object {
              "keys": Array [
                "e",
                "f",
              ],
              "label": "Bar",
            },
          ],
        },
      ]
    }
    innerRef={[Function]}
    items={
      Array [
        Object {
          "a": "1a",
          "b": "2b",
          "c": "3c",
          "d": "4d",
          "e": "5e",
          "f": "6f",
        },
        Object {
          "a": "1a",
          "b": "2b",
          "c": "3c",
          "d": "4d",
          "e": "5e",
          "f": "6f",
        },
      ]
    }
    onResize={[Function]}
    onScroll={[Function]}
  />
</Table__TableWrapper>
`;
